design/cluster-api-provider-metal3/dhcpless_ipam.md

```diff
@@ -0,0 +1,397 @@
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+```
+
+In this flow we have `preprovisioning-pool` which is not associated with any cluster, this is used to provide an IPAddress during
+the pre-provisioning phase as described above.  To reduce the required size of the pool, the IPClaim will be deleted after the
+preprovisioning phase is completed, e.g the BMH resource becomes available.
+
+In the provisioning phase another pool, associated with a cluster is used to template networkData as in the existing process.
+
+#### Assumptions and Open Questions
+
+TODO
+
+### Inspection on initial registration
+
+On initial registration of a host, inspection is triggered immediately but this process cannot complete without preprovisioning network configuration in a DHCP-less environment (because the IPA ramdisk can't connect back to the Ironic API).
+
+To resolve this we can add a new BareMetalHost API `PreprovisioningNetworkDataRequired` which defaults to false, but when set to true will describe that the host cannot move from Registering -> Inspecting until `preprovisioningNetworkDataName` has been set.
+
+An alternative could be to require that the BareMetalHost resources are created with the existing [paused annotation](https://github.com/metal3-io/baremetal-operator/blob/main/docs/api.md#pausing-reconciliation), set to a pre-determined value (e.g `metal3.io/preprovisioning`) which can then be removed by the new controller after `preprovisioningNetworkDataName` has been set, then inspection will be able to succeed.
+
+### Implementation Details/Notes/Constraints
+
+#### IP Reuse
+
+A related issue has been previously addressed via the [IP Reuse](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/ip_reuse.md) functionality - this means we can couple IPClaims to the BareMetalHost resources which will enable consistent IP allocations for pre-provisioning and subsequent provisioning operations (provided the same IPPool is used for both steps)
+
+### Risks and Mitigations
+
+- TODO
+
+### Work Items
+
+TODO
+
+### Dependencies
+
+#### Firstboot agent support
+
+An agent in the IPA ramdisk image is required to consume the network data provided via the processes outlined above.
+
+The Ironic DHCP-less documentation describes using glean (a minimal python-based cloud-init alternative), but we don't
+currently have any community-supported IPA ramdisk image containing this tool.
+
+There are several other options such as cloud-init, or even custom scripts/tooling which may be coupled to the OS, so we
+do not define a specific solution as part of this proposal.
+
+#### Potential config-drive conflict on redeployment
```

Comment:
Thanks @Rozzii - I will move this, as obviously it's a risk but IMO not a blocker to the rest of the proposal, and there are several possible workarounds.

Regarding the workaround you mentioned - does that also require a corresponding change to Ironic so that it sets the disk-label when creating the config-drive image?

Reply:
I have described the two different ISO config drive situations a bit further down https://github.com/metal3-io/metal3-docs/pull/373#discussion_r1465030105, In the first scenario there is no issue, I have done this with glean and glean can remount he IPA boot iso on the running IPA and get the config drive from there (it has a label VMEDIA_BOOT_ISO).

The second scenario is where things don't work out. IT is not a blocker by any means but to support the pre-build IPA ISO use-case we would need to change the redfish driver a bit IMO.

Reply:
Moved to risks and mitigations, and also referenced https://review.opendev.org/c/openstack/ironic-specs/+/906324/1/specs/approved/fix-vmedia-boot-config.rst which was mentioned by @dtantsur as it would be preferable to solve this in the Ironic community rather than in a Metal3 specific way